### PR TITLE
XP Rework

### PR DESCRIPTION
**Pull request by Jagnt**
Increased xp requirements for levels by 10x
Added increasing xp multipliers for higher difficulties
Added flooring for xp numbers (Necessary to avoid massive problems saving data)

### CODE REVIEW

File: gamemodes/horde/gamemode/sv_rank.lua [hunk 1]
```diff
@@ -142,28 +142,28 @@ function HORDE:LoadRank(ply)
...
 
 if GetConVar("horde_enable_sandbox"):GetInt() == 0 and GetConVar("horde_enable_rank"):GetInt() == 1 then
 	hook.Add("Horde_OnEnemyKilled", "Horde_GiveExp", function(victim, killer, wpn)
-
-		local ExpMulti = ExpMultiConvar:GetInt()
 			
-		if HORDE.CurrentDifficulty >= 5 and HORDE.current_wave >= 4 then
-			ExpMulti = 2 * ExpMulti
-		end
+		local WavePercent = Horde.current_wave / 10 --If we ever change the max # of waves this will have to become dynamic instead of 10
```
Comment (wrefgtzweve) [inline gamemodes/horde/gamemode/sv_rank.lua:151]: use `HORDE.max_waves`

File: gamemodes/horde/gamemode/sv_rank.lua [hunk 2]
```diff
@@ -142,28 +142,28 @@ function HORDE:LoadRank(ply)
...
+		local RoundXPMulti = StartXPMult+(WavePercent*(EndXPMult-StartXPMult)) --This gets the xp multi number between min and max multi based on round
```
Comment (wrefgtzweve) [inline gamemodes/horde/gamemode/sv_rank.lua:152]: `EndXPMult-StartXPMult` can be calculated outside the hook

File: gamemodes/horde/gamemode/sv_rank.lua [hunk 3]
```diff
@@ -142,28 +142,28 @@ function HORDE:LoadRank(ply)
...
+		local ExpMulti = RoundXPMulti * ExpMultiConvar:GetInt()
 
 		if HORDE.current_wave <= 0 or GetConVar("sv_cheats"):GetInt() == 1 then return end
 		if killer:IsPlayer() and killer:IsValid() and killer:Horde_GetClass() then
```
Comment (wrefgtzweve) [inline gamemodes/horde/gamemode/sv_rank.lua:156]: shouldathis should return early

File: gamemodes/horde/gamemode/sv_rank.lua [hunk 4]
```diff
@@ -142,28 +142,28 @@ function HORDE:LoadRank(ply)
...
-				killer:Horde_SetExp(class_name, killer:Horde_GetExp(class_name) + (2*ExpMulti) )
+				killer:Horde_SetExp(class_name, killer:Horde_GetExp(class_name) + math.floor(2*ExpMulti) )
 				local p = math.random()
 				if p < 0.01 or (p < 0.1 and killer:Horde_GetGadget() == "gadget_corporate_mindset") then
-                    killer:Horde_AddSkullTokens(1)
+                   			killer:Horde_AddSkullTokens(1)
 				end
 			else
-				killer:Horde_SetExp(class_name, killer:Horde_GetExp(class_name) + (1*ExpMulti) )
+				killer:Horde_SetExp(class_name, killer:Horde_GetExp(class_name) + math.floor(1*ExpMulti) )
```
Comment (wrefgtzweve) [inline gamemodes/horde/gamemode/sv_rank.lua:166]: `Horde_SetExp` should be moved to be used once and the xpmult should be changed like `ExpMulti = ExpMulti * 2` and then be consumed by SetEXP

File: gamemodes/horde/gamemode/sv_rank.lua (hunk 2 above)
Comment (wrefgtzweve) [inline gamemodes/horde/gamemode/sv_rank.lua:152]: use camelCase for all variables

**Review by wrefgtzweve — CHANGES_REQUESTED** [review]

File: gamemodes/horde/gamemode/sv_rank.lua (hunk 4 above)
Comment (Jagnt) [reply gamemodes/horde/gamemode/sv_rank.lua]: consumed lol

File: gamemodes/horde/gamemode/sv_rank.lua (hunk 2 above)
Comment (Jagnt) [reply gamemodes/horde/gamemode/sv_rank.lua]: added, also added camelcase but already resolved it

File: gamemodes/horde/gamemode/sv_rank.lua (hunk 4 above)
Comment (Jagnt) [reply gamemodes/horde/gamemode/sv_rank.lua]: added

File: gamemodes/horde/gamemode/sv_rank.lua (hunk 2 above)
Comment (Jagnt) [reply gamemodes/horde/gamemode/sv_rank.lua]: added

File: gamemodes/horde/gamemode/sv_rank.lua (hunk 1 above)
Comment (Jagnt) [reply gamemodes/horde/gamemode/sv_rank.lua]: added

File: gamemodes/horde/gamemode/sv_rank.lua (hunk 3 above)
Comment (Jagnt) [reply gamemodes/horde/gamemode/sv_rank.lua:157]: you mean set them all to not then return early?

Comment (Jagnt) [reply gamemodes/horde/gamemode/sv_rank.lua]: added

File: gamemodes/horde/gamemode/sh_rank.lua [hunk 5]
```diff
@@ -113,7 +113,7 @@ function plymeta:Horde_SyncExp()
 end
 
 function HORDE:GetExpToNextLevel(level)
-    return math.floor(50 + 2 * math.pow(1.1, level) + level * 25 + 100 * math.floor(level / 5))
+    return math.floor((50 + 2 * math.pow(1.1, level) + level * 25 + 100 * math.floor(level / 5))*10)
```
Comment (wrefgtzweve) [inline gamemodes/horde/gamemode/sh_rank.lua:116]: would make more sense to be ~5 or so, as xp will be significiantly harder to get on all difficulties

Comment (Jagnt) [reply gamemodes/horde/gamemode/sh_rank.lua:116]: i'll set it to 6

**Pull request merged**